### PR TITLE
refactor: remove browser-resolve dependency, inline logic

### PR DESCRIPTION
**Pull request by keithamus**
This PR removes `browser-resolve` as a dependency, inlining the logic steps it makes. It does this for a few reasons:

 1. browser-resolve is very ineficient on disk, it does lots of lookups using its own internal mechanisms where it could rely on `resolve` more.
 2. Similar to point 1 - browser-resolve duplicates a lot of code from `resolve`. Dropping this dependency means shipping less code.
 3. While I _could have_ made a PR to `browser-resolve` to fix the above issues - doing so would reduce `browser-resolve` down to a few LOC, and I'm not sure it is under active maintainence. There are open PRs from a few years ago. I figured it would be best to move the few LOC it needs into this module.

/cc @mislav who has been working with me on this today.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -38,6 +36,17 @@ export default function nodeResolve ( options = {} ) {
...
 
+			if (options.browser && browserMapCache[importer]) {
+				const browser = browserMapCache[importer];
+				if (browser[importee]) {
+					importee = browser[importee];
+				}
+				if (browser[importee] === false) {
+					return ES6_BROWSER_EMPTY;
+				}
+			}
```
Comment (keithamus) [inline src/index.js]: This is the crux of the replacement logic with `browser-resolve`. Given a map from an `importer` - resolve to the values of that key. Here we can also take advantage of returning `ES6_BROWSER_EMPTY` as early as possible - as we don't need to resolbe that beyond what it is doing.

File: src/index.js [hunk 2]
```diff
@@ -69,14 +91,14 @@ export default function nodeResolve ( options = {} ) {
...
 					( err, resolved ) => {
+						if (options.browser && packageBrowserField) browserMapCache[resolved] = packageBrowserField;
+
 						if ( !disregardResult && !err ) {
 							if ( resolved && fs.existsSync( resolved ) ) {
 								resolved = fs.realpathSync( resolved );
 							}
 
-							if ( resolved === COMMONJS_BROWSER_EMPTY ) {
-								fulfil( ES6_BROWSER_EMPTY );
```
Comment (keithamus) [inline src/index.js]: We no longer need this line as we're not relying on `browser-resolve`s empty file.

File: src/index.js [hunk 3]
```diff
@@ -51,13 +60,26 @@ export default function nodeResolve ( options = {} ) {
...
 						basedir: dirname( importer ),
 						packageFilter ( pkg ) {
-							if ( useModule && pkg[ 'module' ] ) {
+							if (options.browser && typeof pkg[ 'browser' ] === 'object') {
+								packageBrowserField = Object.keys(pkg[ 'browser' ]).reduce((browser, key) => {
+									browser[ key ] = pkg[ 'browser' ][key];
+									if (key[0] === '.' && !extname(key)) browser[ key + '.js'] = browser[ key + '.json' ] = browser[ key ];
+									return browser;
+								}, {});
+							}
```
Comment (keithamus) [inline src/index.js]: This is a convenient side-effect-ful mechanism of the `packageFilter` function. We know that if we land inside a `package.json` file, that the resolved `importee` has this mapping inside of its own package.json. Therefore, if we collect these `browser` maps inside of a cache (`browserMapCache`) then by the time we get to the `importee` that owns this package.json being the `importer`, all of it's own dependencies can rely on this cache. 

In other words - it is this operation that collects for us, the mapping to which files are replaced in a browser based environment.

File: src/index.js [hunk 4]
```diff
@@ -51,13 +60,26 @@ export default function nodeResolve ( options = {} ) {
...
+
+							if (options.browser && typeof pkg[ 'browser' ] === 'string') {
+								pkg[ 'main' ] = pkg[ 'browser' ];
+							} else if (options.browser && pkg[ 'browser' ][ pkg[ 'main' ] ]) {
+								pkg[ 'main' ] = pkg[ 'browser' ][ pkg[ 'main' ] ];
```
Comment (keithamus) [inline src/index.js]: This is the second key part of the `browser` key logic - which is that if we hit a module, it's effective `main` field is either `browser` (if it is a string) or `browser[pkg.main]` (if it is an object). Luckily we can use packageFilter exactly how we already are to acheive this.

**Comment by keithamus** [comment]
Tests are failing because Node 6 timed out. https://travis-ci.org/rollup/rollup-plugin-node-resolve/jobs/326432217. Nodes 4 and 9 pass, so I'm sure a rebuild of 6 will pass too.

**Comment by mislav** [comment]
Really happy with how this turned out. 👍 Thanks for the hard work @keithamus!

**Comment by Rich-Harris** [comment]
Thank you! Released 3.0.1

**Comment by Haroenv** [comment]
Hey there, just letting you know that this broke our build, so it might have been different to the original implementation for the browser build: 

https://travis-ci.org/algolia/react-instantsearch/builds/327346207#L686

**Comment by keithamus** [comment]
I'll get a fix for this.

**Comment by keithamus** [comment]
@Haroenv this is now fixed in 3.0.2, thanks to @Rich-Harris 😄

**Comment by Andarist** [comment]
@keithamus this has introduced regression for object browser keys. From what I see they are kinda ignored.

Example:
```js
const importer = '/Users/mateuszburzynski/Desktop/node-uuid/src/v1.js'
const importee = './rng'
const options = { browser: true }

// what gets passed in as arg for packageFilter
const pkg = {"name":"uuid","version":"3.1.0","description":"RFC4122 (v1, v4, and v5) UUIDs","keywords":["uuid","guid","rfc4122"],"license":"MIT","bin":{"uuid":"./bin/uuid"},"devDependencies":{"babel-cli":"^6.26.0","babel-plugin-add-module-exports":"^0.2.1","babel-plugin-transform-es2015-modules-commonjs":"^6.26.0","eslint":"4.5.0","mocha":"3.1.2","runmd":"^1.0.1"},"scripts":{"build":"babel src --out-dir lib","test":"mocha test/test.js","md":"runmd --watch --output=README.md README_js.md","prepare":"npm run build && runmd --output=README.md README_js.md"},"browser":{"./lib/rng.js":"./lib/rng-browser.js","./lib/sha1.js":"./lib/sha1-browser.js","./lib/md5.js":"./lib/md5-browser.js","./src/rng.js":"./src/rng-browser.js","./src/sha1.js":"./src/sha1-browser.js","./src/md5.js":"./src/md5-browser.js"},"repository":{"type":"git","url":"https://github.com/kelektiv/node-uuid.git"},"dependencies":{}}
packageBrowserField {"./lib/rng.js":"./lib/rng-browser.js","./lib/sha1.js":"./lib/sha1-browser.js","./lib/md5.js":"./lib/md5-browser.js","./src/rng.js":"./src/rng-browser.js","./src/sha1.js":"./src/sha1-browser.js","./src/md5.js":"./src/md5-browser.js"}

// result of resolveId call
const resolved = '/Users/mateuszburzynski/Desktop/node-uuid/src/rng.js'
// and this actually should be '/Users/mateuszburzynski/Desktop/node-uuid/src/rng-browser.js'
```

`resolveId` has no knowledge of object browser key - it accept only simple strings in main, so it resolves the whole thing using a simple node algo and the result is not adjusted - especially that importee path was relative to the importer directory and browser field specifies overrides relative to the package root.
